src/amplify-migration/amplifyMigrationScript.ts

```diff
@@ -0,0 +1,289 @@
+      let permalinkLineTrimmed = permalinkLine.trim();
+
+      let hasTrailingQuotation = false;
+      if (permalinkLineTrimmed.endsWith(`"`)) {
+        hasTrailingQuotation = true;
+        permalinkLineTrimmed = permalinkLineTrimmed.slice(0, -1);
+      }
+      const permalinkWithSlash = permalinkLineTrimmed.endsWith("/")
+        ? permalinkLineTrimmed
+        : `${permalinkLineTrimmed}/`;
+      let newPermalink = permalinkWithSlash.toLocaleLowerCase();
+      if (hasTrailingQuotation) {
+        newPermalink = `${newPermalink}"`;
+      }
+      const newFileContent = fileContent.replace(
+        permalinkLine.trim(),
+        newPermalink
+      );
+
+      await fs.promises.writeFile(filePath, newFileContent, "utf-8");
+      //remove trailing slash
+      newPermalink = getRawPermalink(newPermalink);
+      const originalPermalink = getRawPermalink(permalinkLine);
+      if (newPermalink !== originalPermalink) {
+        changedPermalinks[originalPermalink] = newPermalink;
+      }
+      await simpleGit(repoPath).add(filePath);
+    }
+  }
+  console.log(changedPermalinks);
+
+  for (const file of Object.values(mdFiles as {})) {
+    const filePath = path.join(repoPath, file);
+    await changePermalinksInFiles(filePath, repoPath, changedPermalinks);
+  }
+
+  const commitMessage =
+    "chore(Amplify-Migration): Update permalinks in .md files";
+  await simpleGit(repoPath).commit(commitMessage);
+}
+
+/**
+ * Requirements:
+ * convert `permalink: /some/path/` to `some/path`
+ * convert `/some/path` to `some/path`
+ * convert `some/path/` to `some/path`
+ * @param permalink original permalink
+ * @returns raw permalinks without leading/trailing slash
+ */
+function getRawPermalink(permalink: string) {
+  let trimmedPermalink = permalink.trim();
+  if (trimmedPermalink.startsWith(`permalink: `)) {
+    trimmedPermalink = permalink.trim().slice(11);
+  }
+  if (trimmedPermalink.startsWith(`/`)) {
+    trimmedPermalink = trimmedPermalink.slice(1);
+  }
+  if (trimmedPermalink.endsWith("/")) {
+    trimmedPermalink = trimmedPermalink.slice(0, -1);
+  }
+
+  return trimmedPermalink;
+}
+
+async function changePermalinksInFiles(
+  filePath: string,
+  repoPath: string,
+  changedPermalinks: { [key: string]: string }
+) {
+  // read file content
+  let fileContent = await fs.promises.readFile(filePath, "utf-8");
+  let fileChanged = false;
+
+  ({ fileContent, fileChanged } = changeFileContent(
+    fileContent,
+    changedPermalinks
+  ));
+
+  if (fileChanged) {
+    await fs.promises.writeFile(filePath, fileContent, "utf-8");
+    await simpleGit(repoPath).add(filePath);
+  }
+}
+
+function changeFileContent(
+  fileContent: string,
+  changedPermalinks: { [key: string]: string }
+) {
+  let fileChanged = false;
+  // three different regex patterns to take care of
+  // 1. href="original_permalink"
+  // 2. [click here](original_permalink)
+
+  const markdownRegex = /\[(.*?)\]\((.*?)\)/g;
+  const dom = new JSDOM(fileContent);
```

Comment:
> JSDOM assumes that the string is HTML and parses it like so, which is not true in our case (mix of markdown + html); this means that the behaviour here is not well defined, and there exists potential for bugs to arise.

Agreed. My original thought was to parse markdown using markdown-it, then use JSDOM to convert the URLs. 
However, this operation is disruptive, because this would mean that we would be converting all `[blah](some/markdown/format)` to `<a href="some/markdown/format"></a>`. 

This would not be ideal for users who have been only using markdown all this time to edit their pages. 
Additionally, I tried using markdown to parse then get change the urls, but the operation of 
convert markdown -> html -> markdown was not easily achievable. After timeboxing this effort, I stuck to this naive method of parsing the content for markdowns, then changing the permalink from there. 

> we're running a full parse -> construct for each file in this repo, which might be concerning (but given that this is controlled by us, it might be ok)

yes, but just to clarify, we are only changing the `.md` file in the repo. 

My naive hypothesis is that this would cover 80% of all our necessary code changes, have informed ops that there might be some edge cases that this script cannot handle. wdyt?

Reply:
hmm my understanding is that we need to run a search for the original permalink and then detect if it's of the form 

1. `href = permalink` or
2. (desc)[permalink]

and replace it, no?

Reply:
yup, you are right!